### ISSUE TITLE
Processing/Rendering Speed Inaccuracy

### ISSUE THREAD

**Issue by a-barone** [MiSTer-devel/S32X_MiSTer#37]
S32X Core version:
Release 20220607

Hi!

I recorded a couple of videos to show how the S32X core performs side-by-side with a real Genesis 32X setup in attract modes. 
Some games run a bit slower than on real hardware (ex: Star Wars Arcade, Space Harrier), but others have much less slowdown (Mortal Kombat II, Virtua Fighter); so I thought it could be helpful to record such live comparisons.

I'm aware that ultimate accuracy is not the priority right now (though Processing/Rendering speed accuracy is also important for homebrew developers to be able to have in the S32X core a reliable performance reference since the existing emulators are completely unreliable in such aspects) but I believe these comparisons can be useful to understand some other issues as well.

**Real Hardware (Left) | S32X Core (Right)**

**Virtua Fighter** - There's a big performance gap in close-up camera angles; the S32X core is running them much faster than the real hardware.
https://youtu.be/QUrTnOcW5Q4

**Mortal Kombat II** - The title screen animation seems to slow down quite a bit more on real hardware. If you play the game, the Dead Pool stage has quite a different performance, with far fewer performance issues on the S32X Core, especially when hitting the opponent with uppercuts.
https://youtu.be/WS1Z5Zgaw8A

**After Burner Complete** - This is an interesting one. Despite the attract mode gameplay sequences being scripted, the end result seems to differ depending on the rendering speed. 
While the first gameplay sequence goes identical, in the second one (around 2:00) the projectiles seem to hit different objects due to the scrolling speed and aircraft movement speed being different between real hardware and the S32X core; with the attract sequence having a fixed duration.
The attract sequence around 7:15 also shows a bit different partial outcomes; they seem to de-synchronize especially when the game has to render explosions. 
https://youtu.be/Nu2XZxi1oV4
 
**Space Harrier** - Real Hardware is a tad faster with this one; the S32X core seems to be more prone to exhibit a bit of flickering when changing the priority of the sprites to go behind the score table.
https://youtu.be/z27mp21R9ys

**Comment by a-barone** [comment]
S32X Core version:
Release 20220607

Hi!
When I was recording footage for another issue, I noticed that Motocross Championship runs a LOT faster on the S32X core than on real hardware; so I'll leave the recording here for reference:
**Motocross Championship
Real Hardware (Left) | S32X Core (Right)**
https://www.youtube.com/watch?v=0hdru0YKyDs

**Comment by viciious** [comment]
To be honest, I'd leave the current setup as an overclock option :)

**Comment by thehughhefner** [comment]
@a-barone can you try with this build?
[S32X_20221108.zip](https://github.com/MiSTer-devel/S32X_MiSTer/files/9969081/S32X_20221108.zip)

**Comment by a-barone** [comment]
> @a-barone can you try with this build? [S32X_20221108.zip](https://github.com/MiSTer-devel/S32X_MiSTer/files/9969081/S32X_20221108.zip)

Yes, I'll try it later today and report the results today/tomorrow.
Thanks.

**Comment by a-barone** [comment]
> @a-barone can you try with this build? [S32X_20221108.zip](https://github.com/MiSTer-devel/S32X_MiSTer/files/9969081/S32X_20221108.zip)
Motocross Championship
Real Hardware (Left) | S32X Core (Right) (build 20221108)
https://www.youtube.com/watch?v=93r2qUrU2xQ

Virtua Fighter
Real Hardware (Left) | S32X Core (Right) (build 20221108)
https://www.youtube.com/watch?v=MWWkzXucBPY

**Comment by Toryalai1** [comment]
Hi @a-barone 
Could you please revisit this issue ticket? 
Here is the latest unstable s32x core: https://github.com/asturur/S32X_MiSTer/releases/download/unstable-builds/S32X_unstable_20231010_10fda8.rbf

**Comment by a-barone** [comment]
> Hi @a-barone Could you please revisit this issue ticket? Here is the latest unstable s32x core: https://github.com/asturur/S32X_MiSTer/releases/download/unstable-builds/S32X_unstable_20231010_10fda8.rbf

Hi! Yes, but it will take me a week as I'm currently away from home.
I'll post the results here once I test it at home.
Thanks.

**Comment by a-barone** [comment]
Sorry for the long hiatus. I'll try to test it today/tomorrow.

**Comment by asturur** [comment]
@a-barone i made a couple of new builds, you seems to be knowleageble in game and having  hardware to test it, if you want there is a feb 27 build with the saturn code put in without any real science behind it, to try

**Comment by Toryalai1** [comment]
@a-barone did you manage to find time to test this? Please find unstable in the PR that asturur created or the official release

**Comment by asturur** [comment]
@srg320 also made additional changes to the saturn sh7604 so i wonder if in future more fixes will come down the pipeline for this core.

Thanks for taking the time to test again

**Comment by Zet-sensei** [comment]
I tried to backport the SH7604 changes from the Saturn branch to the S32x.

But some BSC changes seem to freeze the core on launch at the moment.
It's going to take a bit of time if I have to search function by function (I'm missing some at the moment) or wait for srg320 to do it by himself.

**Comment by srg320** [comment]
> But some BSC changes seem to freeze the core on launch at the moment.
> It's going to take a bit of time if I have to search function by function (I'm missing some at the moment) or wait for srg320 to do it by himself.

Yes, some tweaks are needed for 16-bit sdram mode in BSC.

**Comment by real-amano** [comment]
Thank you srg320 for updating the sh7604 part of the core.

**Comment by misteraddons** [comment]
[S32X_20260918.zip](https://github.com/user-attachments/files/32391852/S32X_20260918.zip)

Three memory timing fixes, each measured against the SH7604 and 32X hardware manuals.
- Cartridge ROM reads were too fast: Manual requires 9.00 cycles minimum, core did 6.77, now 10.90. Games run code straight from the cart, so this slows everything.
- Frame buffer writes drained too fast: A full 4-word FIFO should cost 5.00 per write, core did 3.22, now 5.13. This slows sustained drawing.
- SDRAM ignored bank-active mode: which the 32X boot ROM enables. Writes should cost 2.00, core did 4.19, now 2.13. This one speeds things up.

Please test: 
- Motocross Championship, Virtua Fighter and Mortal Kombat II ran faster than hardware and should slow toward it
- Kolibri and Spider-Man Web of Fire already ran slow and may get worse
- Also worth checking games still boot and audio is unchanged, since cart reads are on the boot and PWM-DMA paths.

I saw a single glich on Spider-Man Web of Fire - more testing needed.

Related Issues #57 #65

**Comment by srg320** [comment]
If you took measurements using AI and compared them to the manual, that was a bad idea. There are a lot of mistakes in the manual. I took measurements on the original hardware.

A write to SDRAM can take 2 or 4 cycles. That’s normal. It all depends on the previous operation. SH7604 manual doesn’t specify many of the processor’s bus operation characteristics, which were discovered over a long period of time while working on Saturn. These characteristics were identified using a logic analyzer and test programs. 

This is information that the AI can't know because it has not been made public :)
